File: apps/subgraph/src/user-operation.ts
// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// biome-ignore lint/suspicious/noShadowRestrictedNames: <explanation>
// biome-ignore lint/style/useImportType: <explanation>
import { Address, BigInt, Bytes, ethereum, log } from "@graphprotocol/graph-ts";
import { ENTRYPOINT_V060_ADDRESS } from "./const";
import { decodeNonce, decodeNonceKey } from "./nonce";

// -----------------------------------------------------------------------------
// Class
// -----------------------------------------------------------------------------

class UserOperation {
  sender: Address;
  nonceKey: Bytes;
  nonce: BigInt;
  initCode: Bytes;
  callData: Bytes;
  callGasLimit: BigInt;
  verificationGasLimit: BigInt;
  preVerificationGas: BigInt;
  maxFeePerGas: BigInt;
  maxPriorityFeePerGas: BigInt;
  paymasterAndData: Bytes;
  signature: Bytes;
  gasFees: Bytes;

  constructor() {
    this.sender = Address.zero();
    this.nonceKey = new Bytes(0);
    this.nonce = BigInt.zero();
    this.initCode = new Bytes(0);
    this.callData = new Bytes(0);
    this.callGasLimit = BigInt.zero();
    this.verificationGasLimit = BigInt.zero();
    this.preVerificationGas = BigInt.zero();
    this.maxFeePerGas = BigInt.zero();
    this.maxPriorityFeePerGas = BigInt.zero();
    this.paymasterAndData = new Bytes(0);
    this.signature = new Bytes(0);
    this.gasFees = new Bytes(0);
  }
}

class AccountGasLimits {
  callGasLimit: BigInt;
  verificationGasLimit: BigInt;

  constructor(callGasLimit: BigInt, verificationGasLimit: BigInt) {
    this.callGasLimit = callGasLimit;
    this.verificationGasLimit = verificationGasLimit;
  }
}

// -----------------------------------------------------------------------------
// Handlers
// -----------------------------------------------------------------------------

export function handleUserOperationFromCalldata(
  callData: string,
  nonce: BigInt,
  entrypoint: Address,
): UserOperation {
  // Decode the user operation from the input
  log.info("callData: {}", [callData.toString()]);

  // Get the function selector
  const functionSelector = callData.substring(0, 10);
  log.info("functionSelector: {}", [functionSelector.toString()]);

  // Get the function parameters
  const functionParameters = callData.substring(10);
  log.info("functionParameters: {}", [functionParameters.toString()]);

  // Decode the function parameters to hex
  const decodedFunctionParameters = Bytes.fromHexString(functionParameters);
  log.info("decodedFunctionParameters: {}", [
    decodedFunctionParameters.toHexString(),
  ]);

  // Determine the EntryPoint version and set the appropriate decode format
  let decodeFormat: string;
  if (entrypoint.equals(Address.fromString(ENTRYPOINT_V060_ADDRESS))) {
    // v0.6 EntryPoint
    decodeFormat =
      "(address,uint256,bytes,bytes,uint256,uint256,uint256,uint256,uint256,bytes,bytes)[]";
  } else if (
    entrypoint.equals(
      Address.fromString("0x0000000071727de22e5e9d8baf0edac6f37da032"),
    )
  ) {
    // v0.7 EntryPoint
    decodeFormat =
      "(address,uint256,bytes,bytes,bytes32,uint256,bytes32,bytes,bytes)[]";
  } else {
    log.error("Unknown EntryPoint address: {}", [entrypoint.toHexString()]);
    return new UserOperation();
  }

  // Decode the hex function parameters to the user operation params
  const decoded = ethereum.decode(decodeFormat, decodedFunctionParameters);

  // Create a new user operation struct
  const userOperation = new UserOperation();

  // If failed to decode, return empty UserOperation
  if (decoded == null) {
    log.info("Failed to decode user operation params", []);
    return userOperation;
  }

  // Parse the decoded user operation params
  const userOpArray = decoded.toTupleArray<ethereum.Tuple>();

  // Find the matching UserOperation based on nonce
  let matchingUserOp: ethereum.Tuple | null = null;
  for (let i = 0; i < userOpArray.length; i++) {
    // biome-ignore lint/style/useForOf: <explanation>
    // biome-ignore lint/suspicious/noDoubleEquals: <explanation>
    if (userOpArray[i][1].toBigInt() == nonce) {
      matchingUserOp = userOpArray[i];
      break;
    }
  }

  if (matchingUserOp === null) {
    log.info("No matching UserOperation found for nonce: {}", [
      nonce.toString(),
    ]);
    return userOperation;
  }

  // Populate the UserOperation struct based on the EntryPoint version
  if (entrypoint.equals(Address.fromString(ENTRYPOINT_V060_ADDRESS))) {
    // v0.6 EntryPoint
    userOperation.sender = matchingUserOp[0].toAddress();
    userOperation.nonce = matchingUserOp[1].toBigInt();
    userOperation.initCode = matchingUserOp[2].toBytes();
    userOperation.callData = matchingUserOp[3].toBytes();
    userOperation.callGasLimit = matchingUserOp[4].toBigInt();
    userOperation.verificationGasLimit = matchingUserOp[5].toBigInt();
    userOperation.preVerificationGas = matchingUserOp[6].toBigInt();
    userOperation.maxFeePerGas = matchingUserOp[7].toBigInt();
    userOperation.maxPriorityFeePerGas = matchingUserOp[8].toBigInt();
    userOperation.paymasterAndData = matchingUserOp[9].toBytes();
    userOperation.signature = matchingUserOp[10].toBytes();
  } else {
    // v0.7 EntryPoint
    userOperation.sender = matchingUserOp[0].toAddress();

    // nonceKey is 32 bytes hex string
    userOperation.nonceKey = decodeNonceKey(matchingUserOp[1].toBigInt());
    userOperation.nonce = decodeNonce(matchingUserOp[1].toBigInt());

    userOperation.initCode = matchingUserOp[2].toBytes();
    userOperation.callData = matchingUserOp[3].toBytes();

    // accountGasLimits is packed version of `callGasLimit` & `verificationGasLimit`
    const accountGasLimits = matchingUserOp[4].toBytes();
    const gasLimits = unpackAccountGasLimits(accountGasLimits);
    userOperation.callGasLimit = gasLimits.callGasLimit;
    userOperation.verificationGasLimit = gasLimits.verificationGasLimit;

    userOperation.preVerificationGas = matchingUserOp[5].toBigInt();
    userOperation.gasFees = matchingUserOp[6].toBytes();
    userOperation.paymasterAndData = matchingUserOp[7].toBytes();
    userOperation.signature = matchingUserOp[8].toBytes();
  }

  logUserOperation(userOperation, entrypoint);

  return userOperation;
}

// -----------------------------------------------------------------------------
// Utils
// -----------------------------------------------------------------------------

function unpackAccountGasLimits(accountGasLimits: Bytes): AccountGasLimits {
  // biome-ignore lint/suspicious/noDoubleEquals: <explanation>
  if (accountGasLimits.length != 32) {
    // Throw an error if the accountGasLimits is not 32 bytes
    throw new Error(
      `accountGasLimits has unexpected length: ${accountGasLimits.length}`,
    );
  }

  // Unpack callGasLimit (first 16 bytes) and verificationGasLimit (last 16 bytes)
  const callGasLimitBytes = Bytes.fromUint8Array(accountGasLimits.slice(0, 16));
  const verificationGasLimitBytes = Bytes.fromUint8Array(
    accountGasLimits.slice(16, 32),
  );

  const callGasLimit = BigInt.fromUnsignedBytes(callGasLimitBytes);
  const verificationGasLimit = BigInt.fromUnsignedBytes(
    verificationGasLimitBytes,
  );

  log.debug(
    "Unpacked gas limits - callGasLimit: {}, verificationGasLimit: {}",
    [callGasLimit.toString(), verificationGasLimit.toString()],
  );

  return new AccountGasLimits(callGasLimit, verificationGasLimit);
}

function logUserOperation(
  userOperation: UserOperation,
  entrypoint: Address,
): void {
  if (entrypoint.equals(Address.fromString(ENTRYPOINT_V060_ADDRESS))) {
    // v0.6 EntryPoint
    log.info(
      "UserOperation v0.6: sender: {} nonce: {} initCode: {} callData: {} callGasLimit: {} verificationGasLimit: {} preVerificationGas: {} maxFeePerGas: {} maxPriorityFeePerGas: {} paymasterAndData: {} signature: {}",
      [
        userOperation.sender.toHexString(),
        userOperation.nonce.toString(),
        userOperation.initCode.toHexString(),
        userOperation.callData.toHexString(),
        userOperation.callGasLimit.toString(),
        userOperation.verificationGasLimit.toString(),
        userOperation.preVerificationGas.toString(),
        userOperation.maxFeePerGas.toString(),
        userOperation.maxPriorityFeePerGas.toString(),
        userOperation.paymasterAndData.toHexString(),
        userOperation.signature.toHexString(),
      ],
    );
  } else {
    // v0.7 EntryPoint
    log.info(
      "UserOperation v0.7: sender: {} nonce: {} initCode: {} callData: {} callGasLimit: {} verificationGasLimit: {} preVerificationGas: {} gasFees: {} paymasterAndData: {} signature: {}",
      [
        userOperation.sender.toHexString(),
        userOperation.nonce.toString(),
        userOperation.initCode.toHexString(),
        userOperation.callData.toHexString(),
        userOperation.callGasLimit.toString(),
        userOperation.verificationGasLimit.toString(),
        userOperation.preVerificationGas.toString(),
        userOperation.gasFees.toHexString(),
        userOperation.paymasterAndData.toHexString(),
        userOperation.signature.toHexString(),
      ],
    );
  }
}

File: crates/contracts/src/user_operation.rs
// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Copyright 2023-2024 Silius Contributors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// This file is part of Rundler.
//
// Rundler is free software: you can redistribute it and/or modify it under the
// terms of the GNU Lesser General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later version.
//
// Rundler is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with Rundler.
// If not, see https://www.gnu.org/licenses/.

use crate::{tracer::LogInfo, types::UserOperation};
use alloy::{
    dyn_abi::DynSolValue,
    hex,
    primitives::{keccak256, Address, Bytes, FixedBytes, B256, U256},
    rpc::types::RawLog,
};
use core::fmt::Debug;
use eyre::{eyre, Result};
use std::str::FromStr;

// -----------------------------------------------------------------------------
// Utils
// -----------------------------------------------------------------------------

/// A trait for types (events) that can be decoded from a `RawLog`
pub trait EthLogDecode: Send + Sync {
    /// decode from a `RawLog`
    fn decode_log(log: &RawLog) -> Result<Self, Box<dyn std::error::Error>>
    where
        Self: Sized;
}

// From: https://github.com/silius-rs/silius/blob/62cff148f386283bc44114ec9d545eae427489f2/crates/mempool/src/estimate.rs#L80
// License: Apache-2.0

/// Parse the user operation event from the log
/// Since the user operation event is a generic event, we need to parse it with the specific type
/// `T` that implements `EthLogDecode`
pub fn parse_user_op_event<T: EthLogDecode>(event: &LogInfo) -> Result<T> {
    let topics = event
        .topics
        .iter()
        .map(|t| {
            let mut hash_str = t.clone();
            if hash_str.len() % 2 != 0 {
                hash_str.insert(0, '0');
            };
            hex::decode(hash_str).map(|mut b| {
                b.resize(32, 0);
                FixedBytes::<32>::from_slice(b.as_ref())
            })
        })
        .collect::<Result<Vec<FixedBytes<32>>, _>>()
        .map_err(|e| {
            eyre!("simulate handle user op failed on parsing user op event topic hash, {e:?}")
        })?;
    let data = Bytes::from_str(event.data.as_str()).map_err(|e| {
        eyre!("simulate handle user op failed on parsing user op event data: {e:?}")
    })?;

    // Create the log
    let log = RawLog { address: Address::ZERO, topics, data };

    <T>::decode_log(&log)
        .map_err(|err| eyre!("simulate handle user op failed on parsing user op event: {err:?}"))
}

// From: https://github.com/alchemyplatform/rundler/blob/b253c4870b069ffdc16a8ca936fe9ad24e1ac44d/crates/types/src/user_operation.rs#L1-L299
// License: GNU Lesser General Public License v3.0

/// Number of bytes in the fixed size portion of an ABI encoded user operation
const PACKED_USER_OPERATION_FIXED_LEN: usize = 480;

/// Unique identifier for a user operation from a given sender
#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub struct UserOperationId {
    sender: Address,
    nonce: U256,
}

impl UserOperation {
    /// Hash a user operation with the given entry point and chain ID.
    ///
    /// The hash is used to uniquely identify a user operation in the entry point.
    /// It does not include the signature field.
    pub fn op_hash(&self, entry_point: Address, chain_id: u64) -> B256 {
        keccak256(
            DynSolValue::Tuple(vec![
                DynSolValue::FixedBytes(
                    keccak256(self.pack_for_hash()),
                    self.pack_for_hash().len(),
                ),
                DynSolValue::Address(entry_point),
                DynSolValue::Uint(U256::from(chain_id), 256),
            ])
            .abi_encode(),
        )
    }

    /// Get the unique identifier for this user operation from its sender
    pub fn id(&self) -> UserOperationId {
        UserOperationId { sender: self.sender, nonce: self.nonce }
    }

    /// Get the address of the factory entity associated with this user operation, if any
    pub fn factory(&self) -> Option<Address> {
        Self::get_address_from_field(&self.init_code)
    }

    /// Get the address of the paymaster entity associated with this user operation, if any
    pub fn paymaster(&self) -> Option<Address> {
        Self::get_address_from_field(&self.paymaster_and_data)
    }

    /// Extracts an address from the beginning of a data field
    /// Useful to extract the paymaster address from paymaster_and_data
    /// and the factory address from init_code
    pub fn get_address_from_field(data: &Bytes) -> Option<Address> {
        if data.len() < 20 {
            None
        } else {
            Some(Address::from_slice(&data[..20]))
        }
    }

    /// Efficient calculation of the size of a packed user operation
    pub fn abi_encoded_size(&self) -> usize {
        PACKED_USER_OPERATION_FIXED_LEN +
            pad_len(&self.init_code) +
            pad_len(&self.call_data) +
            pad_len(&self.paymaster_and_data) +
            pad_len(&self.signature)
    }

    /// Compute the amount of heap memory the UserOperation takes up.
    pub fn heap_size(&self) -> usize {
        self.init_code.len() +
            self.call_data.len() +
            self.paymaster_and_data.len() +
            self.signature.len()
    }

    /// Gets the byte array representation of the user operation to be used in the signature
    pub fn pack_for_hash(&self) -> Bytes {
        let hash_init_code = keccak256(self.init_code.clone());
        let hash_call_data = keccak256(self.call_data.clone());
        let hash_paymaster_and_data = keccak256(self.paymaster_and_data.clone());

        DynSolValue::Tuple(vec![
            DynSolValue::Address(self.sender),
            DynSolValue::Uint(self.nonce, 256),
            DynSolValue::FixedBytes(hash_init_code, hash_init_code.to_vec().len()),
            DynSolValue::FixedBytes(hash_call_data, hash_call_data.to_vec().len()),
            DynSolValue::Uint(self.call_gas_limit, 256),
            DynSolValue::Uint(self.verification_gas_limit, 256),
            DynSolValue::Uint(self.pre_verification_gas, 256),
            DynSolValue::Uint(self.max_fee_per_gas, 256),
            DynSolValue::Uint(self.max_priority_fee_per_gas, 256),
            DynSolValue::FixedBytes(
                hash_paymaster_and_data,
                hash_paymaster_and_data.to_vec().len(),
            ),
        ])
        .abi_encode()
        .into()
    }
}

/// Calculates the size a byte array padded to the next largest multiple of 32
fn pad_len(b: &Bytes) -> usize {
    (b.len() + 31) & !31
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_hash_zeroed() {
        // Testing a user operation hash against the hash generated by the
        // entrypoint contract getUserOpHash() function with entrypoint address
        // at 0x66a15edcc3b50a663e72f1457ffd49b9ae284ddc and chain ID 1337.
        //
        // UserOperation = {
        //     sender: '0x0000000000000000000000000000000000000000',
        //     nonce: 0,
        //     initCode: '0x',
        //     callData: '0x',
        //     callGasLimit: 0,
        //     verificationGasLimit: 0,
        //     preVerificationGas: 0,
        //     maxFeePerGas: 0,
        //     maxPriorityFeePerGas: 0,
        //     paymasterAndData: '0x',
        //     signature: '0x',
        //   }
        //
        // Hash: 0xdca97c3b49558ab360659f6ead939773be8bf26631e61bb17045bb70dc983b2d
        let operation = UserOperation {
            sender: "0x0000000000000000000000000000000000000000".parse().unwrap(),
            nonce: U256::ZERO,
            init_code: Bytes::default(),
            call_data: Bytes::default(),
            call_gas_limit: U256::ZERO,
            verification_gas_limit: U256::ZERO,
            pre_verification_gas: U256::ZERO,
            max_fee_per_gas: U256::ZERO,
            max_priority_fee_per_gas: U256::ZERO,
            paymaster_and_data: Bytes::default(),
            signature: Bytes::default(),
        };
        let entry_point = "0x66a15edcc3b50a663e72f1457ffd49b9ae284ddc".parse().unwrap();
        let chain_id = 1337;
        let hash = operation.op_hash(entry_point, chain_id);
        assert_eq!(
            format!("0x{:x}", hash),
            "0xdca97c3b49558ab360659f6ead939773be8bf26631e61bb17045bb70dc983b2d"
        );
    }

    #[test]
    fn test_hash() {
        // Testing a user operation hash against the hash generated by the
        // entrypoint contract getUserOpHash() function with entrypoint address
        // at 0x66a15edcc3b50a663e72f1457ffd49b9ae284ddc and chain ID 1337.
        //
        // UserOperation = {
        //     sender: '0x1306b01bc3e4ad202612d3843387e94737673f53',
        //     nonce: 8942,
        //     initCode: '0x6942069420694206942069420694206942069420',
        //     callData: '0x0000000000000000000000000000000000000000080085',
        //     callGasLimit: 10000,
        //     verificationGasLimit: 100000,
        //     preVerificationGas: 100,
        //     maxFeePerGas: 99999,
        //     maxPriorityFeePerGas: 9999999,
        //     paymasterAndData:
        //       '0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef',
        //     signature:
        //       '0xda0929f527cded8d0a1eaf2e8861d7f7e2d8160b7b13942f99dd367df4473a',
        //   }
        //
        // Hash: 0x484add9e4d8c3172d11b5feb6a3cc712280e176d278027cfa02ee396eb28afa1
        let operation = UserOperation {
            sender: "0x1306b01bc3e4ad202612d3843387e94737673f53".parse().unwrap(),
            nonce: U256::from(8942),
            init_code: "0x6942069420694206942069420694206942069420".parse().unwrap(),
            call_data: "0x0000000000000000000000000000000000000000080085".parse().unwrap(),
            call_gas_limit: U256::from(10000),
            verification_gas_limit: U256::from(100000),
            pre_verification_gas: U256::from(100),
            max_fee_per_gas: U256::from(99999),
            max_priority_fee_per_gas: U256::from(9999999),
            paymaster_and_data:
                "0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"
                    .parse()
                    .unwrap(),
            signature: "0xda0929f527cded8d0a1eaf2e8861d7f7e2d8160b7b13942f99dd367df4473a"
                .parse()
                .unwrap(),
        };
        let entry_point = "0x66a15edcc3b50a663e72f1457ffd49b9ae284ddc".parse().unwrap();
        let chain_id = 1337;
        let hash = operation.op_hash(entry_point, chain_id);
        assert_eq!(
            format!("0x{:x}", hash),
            "0x484add9e4d8c3172d11b5feb6a3cc712280e176d278027cfa02ee396eb28afa1"
        );
    }
}
